### ISSUE TITLE
Facebook deprecating oEmbed url

### ISSUE THREAD

**Issue by wgiddens** [wagtail/wagtail#6391]
Facebook is deprecating the oEmbed URLs wagtail uses on October 24, 2020. After that date, the endpoint will return 400 responses.  See: https://developers.facebook.com/docs/plugins/oembed-legacy/

The oEmbed endpoints are being replaced by a new endpoint that requires a registered Facebook app and token.  See: https://developers.facebook.com/docs/plugins/oembed

**Comment by GTpyro** [comment]
Same thing is happening for Instagram: https://developers.facebook.com/docs/instagram/oembed-legacy

**Comment by drimacus182** [comment]
And since yesterday we have dead facebook embeds in production(

**Comment by zerolab** [comment]
@drimacus182 I am sorry to hear that. It is a pity Facebook went down the route of restricted access. While I would've loved for us to stay on top every change a 3rd party does, we do have to carefully balance fixing bugs, new features, roadmap features, and managing issues and review pull requests that the amazing Wagtail community provides.

As someone that uses Facebook embeds in production you are probably the best to help us update the FB embed support.
Or, in the very short term you can fix the problem by creating a custom embed finder as described in https://docs.wagtail.io/en/stable/advanced_topics/embeds.html#custom-embed-finder-classes. The important bit is around how the embed is fetched - https://github.com/wagtail/wagtail/blob/master/wagtail/embeds/finders/oembed.py#L59-L65

You can restrict which providers your custom class applies to via https://docs.wagtail.io/en/stable/advanced_topics/embeds.html#customising-the-provider-list

Hope the above bits of info are helpful in reviving the embeds on your production site.

**Comment by originell** [comment]
If anyone is looking for a quick fix, here's what I implemented. Note that you have  to first create a Facebook App and add the oEmbed Product to it. Below is all the code necessary, if anyone should need a more step-by-step instruction, [I've put that into a blog post](https://www.codista.com/en/blog/wagtail-instagram-new-oembed-api/).

```python
# e.g. in your settings.py:
import os
# ...
OUR_INSTAGRAM_APP_ID = os.environ.get("OUR_INSTAGRAM_APP_ID", "")
OUR_INSTAGRAM_APP_SECRET = os.environ.get("OUR_INSTAGRAM_APP_SECRET", "")
# Bonus point: with the new oembed finder we can also decide to not get the any of the 
# facebook js injected into our site.
OUR_OEMBED_INSTAGRAM_OMITSCRIPT = False
```

then we create a new finder:

```python
# e.g. igram_oembed_finder.py


import json
import re
from urllib import request as urllib_request
from urllib.error import URLError
from urllib.parse import urlencode
from urllib.request import Request

from django.conf import settings
from wagtail.embeds.exceptions import EmbedNotFoundException
from wagtail.embeds.finders.oembed import OEmbedFinder


class InstagramWorkaroundEmbedFinder(OEmbedFinder):
    """A temporary embed finder that supports new Instagram oEmbed Endpoint.

    Sadly Facebook deprecated unauthenticated access to their Instagram oEmbed API.
    Until Wagtail is able to integrate this change into core, this finder can workaround
    this.
    """

    def find_embed(self, url, max_width=None):
        endpoint = self._get_endpoint(url)
        if endpoint != "http://api.instagram.com/oembed":
            return super().find_embed(url, max_width)

        endpoint = "https://graph.facebook.com/v8.0/instagram_oembed"
        # Work out params
        params = self.options.copy()
        params["url"] = url
        params["format"] = "json"
        params[
            "access_token"
        ] = f"{settings.OUR_INSTAGRAM_APP_ID}|{settings.OUR_INSTAGRAM_APP_SECRET}"
        if max_width:
            params["maxwidth"] = max_width
        if getattr(settings, "OUR_OEMBED_INSTAGRAM_OMITSCRIPT", False):
            params["omitscript"] = "true"

        # Perform request
        request = Request(endpoint + "?" + urlencode(params))
        request.add_header("User-agent", "Mozilla/5.0")
        try:
            r = urllib_request.urlopen(request)
        except URLError:
            raise EmbedNotFoundException
        oembed = json.loads(r.read().decode("utf-8"))

        # Convert photos into HTML
        if oembed["type"] == "photo":
            html = '<img src="%s" alt="">' % (oembed["url"],)
        else:
            html = oembed.get("html")

        # Return embed as a dict
        return {
            "title": oembed["title"] if "title" in oembed else "",
            "author_name": oembed["author_name"] if "author_name" in oembed else "",
            "provider_name": oembed["provider_name"]
            if "provider_name" in oembed
            else "",
            "type": oembed["type"],
            "thumbnail_url": oembed.get("thumbnail_url"),
            "width": oembed.get("width"),
            "height": oembed.get("height"),
            "html": html,
        }


embed_finder_class = InstagramWorkaroundEmbedFinder
```

Then we tell Wagtail to use that finder. Again in your settings:

```python
# Wagtail Instagram API Deprecation
WAGTAILEMBEDS_FINDERS = [
    # Handle facebook's public instagram oembed deprecation.
    {"class": "myproject.cms.igram_oembed_finder"}
]
```

That should be it.

**Comment by cnk** [comment]
@originell Thanks for the code (and the blog post). I incorporated your Instagram finder into a PR for putting it in core and then made a similar one for Facebook. 

If you have a chance, could you take a look at https://github.com/wagtail/wagtail/pull/6550 and let me know if it works for you? I don't do a lot on Instagram or Facebook so had a hard time generating realistic examples. I especially had a hard time finding URLs for facebook posts so I am not sure if the list of allowed urls (which I copied from the original oembed finder) are still accurate.

**Comment by originell** [comment]
@cnk sure thing! Will try to take a look tonight

**Comment by originell** [comment]
@cnk I mentioned it in the requested review, but I want to explicitly put it in this thread here as well: thanks a ton for making the time to take care of this! 

I'll make sure to update our blog post once this merged and later when it's released.

**Comment by cnk** [comment]
@originell Thanks so much for the thorough review. I have committed your changes to the PR. We'll see if anyone else has comments while this is in review.

**Comment by npardington** [comment]
I haven't had a chance to test out this branch yet, but would it be possible (or necessary) to catch the specific error codes? Catching the 901 might at least save time.  https://developers.facebook.com/docs/graph-api/reference/oembed-video/#validation-rules

**Comment by cnk** [comment]
@npardington That does seem like a good idea. I'll work on that this evening.

**Comment by gasman** [comment]
Fixed in #6550.
